### DISCUSSION TITLE
Welcome to emacs-application-framework Discussions!

### DISCUSSION THREAD

**Discussion by manateelazycat** [emacs-eaf/emacs-application-framework#467] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by manateelazycat** [comment]
If you need help with something or have some cool ideas to share, please post it here instead of the issue, this is more like forum page for EAF! ;)

**Comment by Young-people** [comment]
Does it work on CentOS 7? How to install dependency? I want to use a browser.

**Comment by sch1z00** [comment]
wow! now I can browse [ZeroNet](https://zeronet.io) in Emacs :)

**Reply by sch1z00** [reply]
writing inside zeronet forms don't work, keys are interpreted as eaf-browser keybinding

**Reply by manateelazycat** [reply]
> writing inside zeronet forms don't work, keys are interpreted as eaf-browser keybinding

we have whitelist fix this problem，can you give details url？thanks

**Reply by sch1z00** [reply]
> > writing inside zeronet forms don't work, keys are interpreted as eaf-browser keybinding
> 
> we have whitelist fix this problem，can you give details url？thanks

in every page (forum, blog, microblog...) where you can input text there's the same problem

if you don't have a zeronet installation you can access through proxy and try "Filter: site name" on the top-left side of the first aggregative page

https://zeronet.now.im/

an introduction on zeronet structure is on ZeroNet Developer Documentation » https://zeronet.io/docs/site_development/getting_started/

**Comment by dalanicolai** [comment]
Hi guys! I could not find how to visit a new url within the same tab, i.e. without opening a new tab. `eaf-open-browser` always opens a new tab (if the new url does not match a url of a currently open buffer). You have an idea?

**Reply by manateelazycat** [reply]
> Hi guys! I could not find how to visit a new url within the same tab, i.e. without opening a new tab. `eaf-open-browser` always opens a new tab (if the new url does not match a url of a currently open buffer). You have an idea?

open_url in core/browser.el

**Reply by dalanicolai** [reply]
Ah, I guess you were mistaking. The command I was looking for is `(insert_or_)edit_url`. Don't know how I overlooked that one. Anyway, thanks to your hint I finally found it.

**Comment by dalanicolai** [comment]
Should the pdf-viewer work with djvu? Because here, although the eaf pdf-viewer opens djvu (and the the extension is in the default `eaf-pdf-extension-list`), it does not show anything.

**Reply by MatthewZMD** [reply]
@luhuaei any ideas?

**Reply by luhuaei** [reply]
> @luhuaei any ideas?

I see the `pymupdf` support extensions don't has `djvu`, and use `mupdf` open `djvu` file failed.  

So use `pdf-viewer` open `djvu` file will not show anything. @dalanicolai

**Reply by MatthewZMD** [reply]
Sounds good, I've removed djvu from eaf-pdf-extension-list to prevent future confusion.

**Comment by dalanicolai** [comment]
Hello again!

I am a Spacemacs user, and I have noticed that since the witch to `epc`, `ESC` does not work anymore to exit from insert state (e.g. in the browser). Can you (help me to) fix that?

**Reply by dalanicolai** [reply]
Okay, so this part is fixed by not requiring `eaf-evil` (as suggested [here](https://github.com/manateelazycat/emacs-application-framework/wiki/Evil))

**Comment by dalanicolai** [comment]
Also beside the `SPC`-leader key, Spacemacs uses `,` as a major-mode leader key.

I had added functionality to emulate the leader key in EAF, similar to [how you achieve that for SPC](https://github.com/manateelazycat/emacs-application-framework/wiki/Evil).
So the code looks like this:
```
      (setq eaf-evil-leader-for-major-keymap (make-sparse-keymap))
      (define-key eaf-evil-leader-for-major-keymap (kbd "h") 'eaf-open-browser-with-history)
      (define-key eaf-evil-leader-for-major-keymap (kbd "d") 'eaf-toggle-dark-mode)
      (define-key eaf-evil-leader-for-major-keymap (kbd "s") 'eaf-search-it)
      (add-hook 'evil-normal-state-entry-hook
                (lambda ()
                  (when (derived-mode-p 'eaf-mode)
                    (define-key eaf-mode-map (kbd "C-,") eaf-evil-leader-for-major-keymap))))

      (define-key key-translation-map (kbd ",")
        (lambda (prompt)
          (if (derived-mode-p 'eaf-mode)
              (if (and (string= eaf--buffer-app-name "browser")
                       (eaf-call-sync "call_function" eaf--buffer-id "is_focus"))
                  (kbd ",")
                (kbd "C-,"))))))))
```
This worked when using dbus (when I used just eaf-call without the -sync). However, currently when I use `describe-key` for `,` in an EAF app, it tells me it is bound to `eaf-send-key`.

So I hope you could (help me also with) get this functionality back. :pray:

**Reply by dalanicolai** [reply]
Ah, okay I found this problem also. I see you have changed the condition in the if clause:
```
  ("browser" (if (string= (eaf-call-sync "call_function" eaf--buffer-id "is_focus") "True")
                           (kbd "SPC")
                         (kbd eaf-evil-leader-key)))
```

So by changing mine similarly, it does work now.

**Comment by nanarayanan** [comment]
There is a dependency on f.el (http://github.com/rejeep/f.el) since the following commit
```
commit 81b59fc02f9c2f3b74afb63e096af448a9a5b06f
Author: wcq062821 <wcq-062821@163.com>
Date:   Tue Jun 1 23:58:01 2021 +0800

    add eaf-open-pdf-with-history function (#690)
```
as it has introduced a call to `(f-write-text "" 'utf-8 pdf-history-file-path)`
So there is a need to download and install `f.el` from melpa (or other repository).

**Reply by MatthewZMD** [reply]
Thanks, this has been addressed in the newest commit, I've removed this dependency.

**Comment by nanarayanan** [comment]
install_eaf.sh on Ubuntu failed with the error: `./install-eaf.sh: 9: Syntax error: "(" unexpected   ` 

The error was due to `POSITIONAL=()` at line number 9 in `install-eaf.sh`
This was seen in the commit 
```
commit 4b5d4294333c973ab3806ef891a95ecb9e60f556 (HEAD -> master, origin/master, origin/HEAD)
Date:   Fri Jul 23 16:59:30 2021 -0400
```
Changing the shell from `sh` to `bash` helped fix it:
```
--- a/install-eaf.sh                                                                                                                              
+++ b/install-eaf.sh                                                                                                                              
@@ -1,4 +1,4 @@                                                                                                                                   
-#!/usr/bin/env sh                                                                                                                                
+#!/usr/bin/env bash 
```

**Reply by MatthewZMD** [reply]
My bad, I forgot that it was bash-only syntax, the latest version has it fixed.

**Comment by jzxu** [comment]
Hi, thanks for writing EAF! I'm using the browser and pdf viewer and they are working really well! The only problem I have experienced so far is that the browser buffer sometimes stops responds to any keybindings at all. Even C-x and C-c do not work. I have to click on the modeline with my mouse in order to get out of this mysterious state, and then the keybindings start working again. Is this a bug or a feature that I don't understand?

Thanks!

**Reply by manateelazycat** [reply]
> Hi, thanks for writing EAF! I'm using the browser and pdf viewer and they are working really well! The only problem I have experienced so far is that the browser buffer sometimes stops responds to any keybindings at all. Even C-x and C-c do not work. I have to click on the modeline with my mouse in order to get out of this mysterious state, and then the keybindings start working again. Is this a bug or a feature that I don't understand?
> 
> Thanks!

https://github.com/emacs-eaf/emacs-application-framework#why-doesnt-eaf-receive-input-events-on-linux-window-manager

**Reply by jzxu** [reply]
Thanks for the fast reply! That fixed the problem! Just FYI my "wmctrl -m" output is "Xfwm4".

**Reply by manateelazycat** [reply]
> Thanks for the fast reply! That fixed the problem! Just FYI my "wmctrl -m" output is "Xfwm4".

you're welcome! :)

**Reply by jzxu** [reply]
I'm experiencing another strange issue with eaf-browser. Whenever my mouse cursor is over a browser window and I press any key, even just the control key and nothing else, the mouse cursor will move to the bottom right corner of my screen.

- This happens even if the browser window is not the active window.
- This doesn't happen if my mouse cursor is over any other window, even if the browser window is active.

I'm guessing this is related to my previous problem of x event handling.

Thanks for your help again!

**Reply by manateelazycat** [reply]
> I'm experiencing another strange issue with eaf-browser. Whenever my mouse cursor is over a browser window and I press any key, even just the control key and nothing else, the mouse cursor will move to the bottom right corner of my screen.
> 
> * This happens even if the browser window is not the active window.
> * This doesn't happen if my mouse cursor is over any other window, even if the browser window is active.
> 
> I'm guessing this is related to my previous problem of x event handling.
> 
> Thanks for your help again!

yes，cause by preview setup, it's use workaround to fix key press issue in xfwm

**Reply by jzxu** [reply]
Is there some way to fix this? Will it be resolved if I switch to a different desktop environment such as Gnome?

**Reply by manateelazycat** [reply]
> Is there some way to fix this? Will it be resolved if I switch to a different desktop environment such as Gnome?

yes，gnome or kde won't move cursor

**Reply by jzxu** [reply]
I switched to KDE and still experience the same problem.  My "wmctrl -m" output is now "KWin". I had to add this to the "eaf-wm-focus-fix-wms" variable, otherwise it still wouldn't recognize any keyboard events once I click the browser buffer. I'm using Debian, KDE Plasma 5.21.5 and KDE Frameworks 5.86.0.

**Reply by manateelazycat** [reply]
> I switched to KDE and still experience the same problem. My "wmctrl -m" output is now "KWin". I had to add this to the "eaf-wm-focus-fix-wms" variable, otherwise it still wouldn't recognize any keyboard events once I click the browser buffer. I'm using Debian, KDE Plasma 5.21.5 and KDE Frameworks 5.86.0.

Gnome and KDE don't need move cursor, maybe something daemon serve is running in your system? 

Can you install clean KDE for test? 

I use KDE before, works well.

**Reply by jzxu** [reply]
It was a clean KDE install, as far as I could tell. I don't use KDE and it was the first time I installed it on my computer. Any suggestions on how I can debug this problem?

Thanks!

**Reply by manateelazycat** [reply]
> It was a clean KDE install, as far as I could tell. I don't use KDE and it was the first time I installed it on my computer. Any suggestions on how I can debug this problem?
> 
> Thanks!

You can uncomment this line https://github.com/emacs-eaf/emacs-application-framework/blob/249b6c9238eaad7bf3a91d728367e8f0fa8b930b/core/view.py#L111 to debug what's wrong. ;)

**Reply by MatthewZMD** [reply]
> It was a clean KDE install, as far as I could tell. I don't use KDE and it was the first time I installed it on my computer. Any suggestions on how I can debug this problem?
> 
> Thanks!

I actively use KDE and never encounter this problem. Hmmmm

**Reply by jzxu** [reply]
> You can uncomment this line
> 
> https://github.com/emacs-eaf/emacs-application-framework/blob/249b6c9238eaad7bf3a91d728367e8f0fa8b930b/core/view.py#L111
> 
> to debug what's wrong. ;)

Sorry for the radio silence. I tried uncommenting that line and reinstalling eaf. However, I don't see any output in the terminal I launched emacs from. Where is the output supposed to show up?

**Reply by manateelazycat** [reply]
> > You can uncomment this line
> > https://github.com/emacs-eaf/emacs-application-framework/blob/249b6c9238eaad7bf3a91d728367e8f0fa8b930b/core/view.py#L111
> > 
> > to debug what's wrong. ;)
> 
> Sorry for the radio silence. I tried uncommenting that line and reinstalling eaf. However, I don't see any output in the terminal I launched emacs from. Where is the output supposed to show up?

```*eaf*``` buffer

**Reply by jzxu** [reply]
Thanks for your continued help so far, Andy and Matthew! I've spent a long time debugging and haven't gotten anywhere. Since both XFCE and KDE are listed as supported window managers, I don't understand why I can't get the keyboard focus to work without using the eaf-wm-focus-fix-wms workaround (the mouse warping behavior of this workaround is IMHO more annoying than not getting keyboard events at all). I wonder if it has something to do with the versions of the libraries I am using?

pyqt5: 5.15.4
xserver: 1.20.11
XFCE: 4.16
KDE: 5:112

Just for reference I'm also including what I did for debugging, in case it is useful somehow.

I used "xwininfo -tree" to print all windows associated with emacs. I managed to find the window ids for the EAF browser window and the normal emacs window this way.

I used "xev -id <window id> -event keyboard" to monitor the keyboard events for the emacs window and the EAF window. I confirmed that I do indeed see the keyboard events for each window this way. It also seems like the id of the EAF window changes every once in a while.

What I observe in this setup is that I don't even need to click on the EAF window for the focus to switch to it. I simply have to hover my mouse over the EAF window, and then all keyboard events will go to it. If I then move the mouse back to a regular emacs window, emacs starts getting keyboard events again. I guess you guys are already aware of this as it's the basis for the mouse warping work-around.

I tried to change the line "(shell-command-to-string (format "wmctrl -i -a $(wmctrl -lp | awk '$3==%s {print $1; exit}')" (emacs-pid)))" in eaf.el to several other approaches to change window focus. I first passed an explicit window id to "wmctrl -i -a" to rule out any possibility that the window id wasn't computed correctly. I also tried "xdotool windowactivate" and "xdotool windowfocus" with explicit window ids. None of this worked.

**Reply by jzxu** [reply]
I finally figured out why my emacs was never capturing keyboard events when my mouse cursor is over an EAF buffer. It had nothing to do with my window manager. My emacs is compiled from the git repo and I had used the option "--with-x-toolkit=lucid" in the configure script. Once I removed that option (which means using the default gtk3 toolkit) and recompiled, EAF started working as expected. I don't have to use the eaf-wm-focus-fix-wms workaround either.

My guess is that EAF probably won't work with anything other than --with-x-toolkit=gtk3? You can actually check for this using the elisp variable "system-configuration-features". It will contain the string "GTK3" if using gtk3 x-toolkit, otherwise it will contain "LUCID" or "ATHENA". It would be good to warn the user in the latter case so that they won't go on a wild goose chase like I did.

**Reply by manateelazycat** [reply]
> I finally figured out why my emacs was never capturing keyboard events when my mouse cursor is over an EAF buffer. It had nothing to do with my window manager. My emacs is compiled from the git repo and I had used the option "--with-x-toolkit=lucid" in the configure script. Once I removed that option (which means using the default gtk3 toolkit) and recompiled, EAF started working as expected. I don't have to use the eaf-wm-focus-fix-wms workaround either.
> 
> My guess is that EAF probably won't work with anything other than --with-x-toolkit=gtk3? You can actually check for this using the elisp variable "system-configuration-features". It will contain the string "GTK3" if using gtk3 x-toolkit, otherwise it will contain "LUCID" or "ATHENA". It would be good to warn the user in the latter case so that they won't go on a wild goose chase like I did.

Thank you for your help, I have pushed the patch: https://github.com/emacs-eaf/emacs-application-framework/commit/0e4074726acb2c95b8ac3335ebc044867d256501 to notify user.

**Comment by nbfields** [comment]
Hi! I'm a relatively new user to Emacs, having migrated over from Neovim a few months ago. I quickly took to seeing how many of my old customizations I in Neovim I could bring over to Emacs, with Evil-mode and things associated with it. I can safely say that I've fallen head over heels for Emacs and can't see myself going back to Neovim any time soon. EAF is incredible (though I have some questions/issues(?) to report, which is part of why I came here) and I am excited beyond words to see where this amazing piece of software goes next.

**Reply by MatthewZMD** [reply]
Awesome, glad you're enjoying EAF. Feel free to ask questions here or open an issue

**Comment by rdiaz02** [comment]
Hello, I am not sure how to modify `eaf-mode-map*`.  I want my setting for `M-o` not be touched by EAF. So I have, in my init.el 

```
 (setq eaf-mode-map*
		  (let ((map (make-sparse-keymap)))
		    (define-key map (kbd "M-o") #'ace-window)
		    (define-key map (kbd "C-h m") #'eaf-describe-bindings)
...
)
```

However, under eaf-browser, `M-o` still gives `eaf-py-proxy-eval_js`. 

 I've checked and `C-h v eaf-mode-map*` correctly shows my binding:

```
eaf-mode-map* is a variable defined in ‘eaf.el’.

Its value is shown below.

Keymap for default bindings available in all apps.

Value:
(keymap
 (return . eaf-send-key)
 (backspace . eaf-send-key)
...
 (27 keymap
     (91 . eaf-share-path-or-url)
     (47 . eaf-get-path-or-url)
     (39 . eaf-toggle-fullscreen)
     (111 . ace-window)))


```

**Reply by manateelazycat** [reply]
Use eaf-bind-key like my example: https://github.com/manateelazycat/lazycat-emacs/blob/6115133f822afb44268578f4971919edd2cc63eb/site-lisp/config/init-eaf.el#L121

**Reply by rdiaz02** [reply]
Thanks,  that worked perfectly!!
